File: Source/Rewards/Challenges/ChainedTogether.cs
using System;
using System.Collections.Generic;
using WCSharp.Api;
using static WCSharp.Api.Common;

public static class ChainedTogether
{
    private const int REQUIRED_PLAYERS = 2;
    private static Dictionary<string, Chain> KittyLightnings = new Dictionary<string, Chain>();
    private static Dictionary<string, List<Kitty>> kittyGroups = new Dictionary<string, List<Kitty>>();
    private static float timerInterval = 0.01f;
    private static Random rng = Globals.RANDOM_GEN;
    private static timer MoveChainTimer;
    private static bool EventTriggered { get; set; } = false;
    private static bool EventStarted { get; set; } = false;
    private static bool IsStartingConditionValid = true;

    private static void TriggerEventOnLastSafeZone()
    {
        if (Gamemode.CurrentGameMode != GameMode.Standard) return; // Only occurs in Standard Gamemode.
        if (EventStarted || EventTriggered) return; // Don't trigger multiple times.
        if (!IsStartingConditionValid) return;

        List<Kitty> allKitties = Globals.ALL_KITTIES_LIST;

        if (allKitties.Count < REQUIRED_PLAYERS) return; // Need at least 2 players to trigger event.

        for (int i = 0; i < allKitties.Count - 1; i++)
        {
            if (!IsInLastSafezone(allKitties[i]))
            {
                return;
            }
        }

        TriggerEvent();
    }

    public static void TriggerEvent()
    {
        Utility.TimedTextToAllPlayers(4.0f, $"{Colors.COLOR_TURQUOISE}Chained Together Event Requirements Complete!{Colors.COLOR_RESET} {Colors.COLOR_YELLOW}Activating next round!{Colors.COLOR_RESET}");
        EventTriggered = true;
    }
    private static void UpdateStartingCondition(Kitty kitty)
    {
        int currentSafezone = kitty.CurrentSafeZone;
        List<Kitty> kitties = Globals.ALL_KITTIES_LIST;
        bool skippedSafezone = false;

        for (int i = 0; i < kitties.Count; i++)
        {
            Kitty currentKitty = kitties[i];
            if (currentKitty.CurrentSafeZone != currentSafezone - 1 && currentKitty.CurrentSafeZone != currentSafezone)
            {
                skippedSafezone = true;
                break;
            }
        }

        if (skippedSafezone)
        {
            IsStartingConditionValid = false;
        }
    }

    public static void StartEvent()
    {
        if (!EventTriggered)
        {
            IsStartingConditionValid = true;
            return;
        }

        EventStarted = true;

        try
        {
            SetGroups();
            MoveChainTimer ??= CreateTimer();
            TimerStart(MoveChainTimer, timerInterval, true, MoveChain);
        }
        catch (Exception e)
        {
            Logger.Warning($"Error in ChainedTogether.StartEvent {e.Message}");
            throw;
        }
    }

    private static void MoveChain()
    {
        foreach (var chain in KittyLightnings.Values)
        {
            chain.Constrain();
        }
    }
    
    // Unused anyway
    private static void RegenerateGroup(string kittyName)
    {
        if (!kittyGroups.TryGetValue(kittyName, out var group)) return;
        group.RemoveAll(k => k.Name == kittyName);
        RechainGroup(group);
    }

    private static void FreeKittiesFromGroup(string kittyName, bool isVictory = false)
    {
        if (!kittyGroups.TryGetValue(kittyName, out var group)) return;
        foreach (var k in group)
        {
            k.IsChained = false;
            if (KittyLightnings.TryGetValue(k.Name, out var c))
            {
                c.Dispose();
                KittyLightnings.Remove(k.Name);
            }
            if (isVictory) AwardChainedTogether(k);
        }
        foreach (var k in group) kittyGroups.Remove(k.Name);
    }

    private static void SetGroups()
    {
        var all = new List<Kitty>(Globals.ALL_KITTIES_LIST);
        int count = all.Count;
        for (int i = count - 1; i > 0; i--)
        {
            int j = rng.Next(i + 1);
            var temp = all[i];
            all[i] = all[j];
            all[j] = temp;
        }

        List<List<Kitty>> groups = new();
        if (count < 3)
            groups.Add(all);
        else
        {
            int groupsOfThree = count / 3;
            int rem = count % 3;
            if (rem == 1)
            {
                groupsOfThree--;
                rem += 3;
            }
            int idx = 0;
            for (int i = 0; i < groupsOfThree; i++)
            {
                groups.Add(all.GetRange(idx, 3));
                idx += 3;
            }
            if (idx < count)
                groups.Add(all.GetRange(idx, count - idx));
        }

        foreach (var g in groups)
            RechainGroup(g);
    }

    private static void RechainGroup(List<Kitty> group)
    {
        for (int i = 0; i < group.Count - 1; i++)
        {
            var a = group[i];
            var b = group[i + 1];
            a.IsChained = b.IsChained = true;
            var chain = ObjectPool.GetEmptyObject<Chain>();
            chain.SetKitties(a, b);
            KittyLightnings[a.Name] = chain;
            kittyGroups[a.Name] = group;
            kittyGroups[b.Name] = group;
        }
    }

    public static void ReachedSafezone(Kitty kitty)
    {
        try
        {
            if (!EventStarted)
            {
                UpdateStartingCondition(kitty);
                if (IsInLastSafezone(kitty))
                    TriggerEventOnLastSafeZone();
                return;
            }
            if (!IsInLastSafezone(kitty)) return;
            FreeKittiesFromGroup(kitty.Name, true);
        }
        catch (Exception e)
        {
            Logger.Warning($"Error in ChainedTogether.ReachedSafezone {e.Message}");
            throw;
        }
    }

    private static void AwardChainedTogether(Kitty kitty)
    {
        Utility.CreateSimpleTextTag($"{Colors.COLOR_RED}Chained Together!{Colors.COLOR_RESET}", 2.0f, kitty.Unit);

        DifficultyLevel level = (DifficultyLevel)Difficulty.DifficultyValue;

        var awards = Globals.GAME_AWARDS_SORTED.Auras;
        string nameOfAward;

        switch (level)
        {
            case >= DifficultyLevel.Nightmare:
                nameOfAward = nameof(awards.ChainedNightmareAura);
                break;
            case >= DifficultyLevel.Impossible:
                nameOfAward = nameof(awards.ChainedImpossibleAura);
                break;
            case >= DifficultyLevel.Hard:
                nameOfAward = nameof(awards.ChainedHardAura);
                break;
            default:
                nameOfAward = nameof(awards.ChainedNormalAura);
                break;
        }

        AwardManager.GiveReward(kitty.Player, nameOfAward);
    }

    private static bool IsInLastSafezone(Kitty kitty)
    {
        // Check if the checks made with this function can be removed by calling the
        // functions TriggerEventOnLastSafeZone and ReachedSafezone on Source\Events\VictoryZone\FinalSafezone.cs
        // I'm afraid there could be a race condition by doing so
        return kitty.CurrentSafeZone == RegionList.SafeZones.Length - 1;
    }
}

public class Chain
{
    public Kitty FirstKitty { get; private set; }
    public Kitty SecondKitty { get; private set; }
    public lightning Lightning { get; private set; }

    private static readonly Dictionary<DifficultyLevel, (int good, int far, int max)> Ranges = new()
    {
        { DifficultyLevel.Normal,     (600, 700, 800) },
        { DifficultyLevel.Hard,       (650, 750, 850) },
        { DifficultyLevel.Impossible, (700, 800, 900) },
        { DifficultyLevel.Nightmare,  (750, 850, 950) }
    };

    public void SetKitties(Kitty a, Kitty b)
    {
        FirstKitty = a;
        SecondKitty = b;
        Lightning?.Dispose();
        Lightning = AddLightning("WHCH", true,
            a.Unit.X, a.Unit.Y, b.Unit.X, b.Unit.Y);
    }

    public void Constrain()
    {
        float x1 = FirstKitty.Unit.X, y1 = FirstKitty.Unit.Y;
        float x2 = SecondKitty.Unit.X, y2 = SecondKitty.Unit.Y;
        float dx = x2 - x1, dy = y2 - y1;
        float dist = (float)Math.Sqrt(dx * dx + dy * dy);
        int max = GetRange("max");

        if (dist > max && dist > 0f)
        {
            float excess = dist - max;
            float nx = dx / dist, ny = dy / dist;
            bool firstCanMove = FirstKitty.Alive && !FirstKitty.Unit.IsPaused;
            bool secondCanMove = SecondKitty.Alive && !SecondKitty.Unit.IsPaused;
            float w1 = firstCanMove ? (secondCanMove ? 0.5f : 1f) : 0f;
            float w2 = secondCanMove ? (firstCanMove ? 0.5f : 1f) : 0f;

            // Move FirstKitty with terrain check
            float newX1 = nx * excess * w1;
            float newY1 = ny * excess * w1;

            if (IsTerrainPathable(FirstKitty.Unit.X + newX1, y1, PATHING_TYPE_WALKABILITY))
            {
                newX1 = 0;
            }

            if (IsTerrainPathable(x1, FirstKitty.Unit.Y + newY1, PATHING_TYPE_WALKABILITY))
            {
                newY1 = 0;
            }
          
            FirstKitty.Unit.SetPathing(false);
            FirstKitty.Unit.X += newX1;
            FirstKitty.Unit.Y += newY1;
            FirstKitty.Unit.SetPathing(true);

            // Move SecondKitty with terrain check
            float newX2 = nx * excess * w2;
            float newY2 = ny * excess * w2;
            if (IsTerrainPathable(SecondKitty.Unit.X + newX2, y2, PATHING_TYPE_WALKABILITY))
            {
                newX2 = 0;
            }

            if (IsTerrainPathable(x2, SecondKitty.Unit.Y + newY2, PATHING_TYPE_WALKABILITY))
            {
                newY2 = 0;
            }

            SecondKitty.Unit.SetPathing(false);
            SecondKitty.Unit.X -= newX2;
            SecondKitty.Unit.Y -= newY2;
            SecondKitty.Unit.SetPathing(true);
        }

        // update lightning and color
        MoveLightning(Lightning, true,
            FirstKitty.Unit.X, FirstKitty.Unit.Y,
            SecondKitty.Unit.X, SecondKitty.Unit.Y);
        UpdateColor(dist);
    }


    private void UpdateColor(float dist)
    {
        int good = GetRange("good"), far = GetRange("far");
        float r = 0f, g = 1f, b = 0f, a = 1f;
        if (dist > far) { r = 1f; g = 0f; b = 0f; }
        else if (dist > good) { r = 1f; g = 1f; b = 0f; }
        SetLightningColor(Lightning, r, g, b, a);
    }

    private static int GetRange(string type)
    {
        DifficultyLevel lvl = (DifficultyLevel)Difficulty.DifficultyValue;
        var key = lvl >= DifficultyLevel.Nightmare
            ? DifficultyLevel.Nightmare
            : lvl >= DifficultyLevel.Impossible
                ? DifficultyLevel.Impossible
                : lvl >= DifficultyLevel.Hard
                    ? DifficultyLevel.Hard
                    : DifficultyLevel.Normal;
        var tuple = Ranges[key];
        return type switch
        {
            "good" => tuple.good,
            "far" => tuple.far,
            "max" => tuple.max,
            _ => throw new ArgumentException($"Invalid range type {type}")
        };
    }

    public void Dispose()
    {
        Lightning?.Dispose();
        FirstKitty.IsChained = false;
        SecondKitty.IsChained = false;
        ObjectPool.ReturnObject(this);
    }
}

File: src/Rewards/Challenges/ChainedTogether.ts
import { Logger } from 'src/Events/Logger/Logger'
import { Kitty } from 'src/Game/Entities/Kitty/Kitty'
import { CurrentGameMode } from 'src/Gamemodes/CurrentGameMode'
import { GameMode } from 'src/Gamemodes/GameModeEnum'
import { Globals } from 'src/Global/Globals'
import { RegionList } from 'src/Global/RegionList'
import { Difficulty } from 'src/Init/Difficulty/Difficulty'
import { DifficultyLevel } from 'src/Init/Difficulty/DifficultyOption'
import { Colors } from 'src/Utility/Colors/Colors'
import { MemoryHandler } from 'src/Utility/MemoryHandler/MemoryHandler'
import { Utility } from 'src/Utility/Utility'
import { Timer } from 'w3ts'
import { AwardManager } from '../Rewards/AwardManager'

export class ChainedTogether {
    private static REQUIRED_PLAYERS = 2
    private static KittyLightnings: Map<string, Chain> = new Map()
    private static kittyGroups: Kitty[][] = [] // Convert this into a dictionary
    private static timerInterval = 0.1
    private static MoveChainTimer: Timer
    private static EventTriggered: boolean = false
    private static EventStarted: boolean = false
    private static IsStartingContidionValid: boolean = true

    private static TriggerEventOnLastSafeZone = () => {
        if (CurrentGameMode.active !== GameMode.Standard) return // Only occurs in Standard Gamemode.
        if (ChainedTogether.EventStarted || ChainedTogether.EventTriggered) return // Don't trigger multiple times.
        if (!ChainedTogether.IsStartingContidionValid) return

        const allKitties = Globals.ALL_KITTIES_LIST

        if (allKitties.length < ChainedTogether.REQUIRED_PLAYERS) return // Need at least 2 players to trigger event.

        for (let i = 0; i < allKitties.length - 1; i++) {
            if (!ChainedTogether.IsInLastSafezone(allKitties[i])) {
                return
            }
        }

        ChainedTogether.TriggerEvent()
    }

    public static TriggerEvent = () => {
        Utility.TimedTextToAllPlayers(
            4.0,
            `${Colors.COLOR_TURQUOISE}Chained Together Event Requirements Complete!${Colors.COLOR_RESET} ${Colors.COLOR_YELLOW}Activating next round!${Colors.COLOR_RESET}`
        )
        ChainedTogether.EventTriggered = true
    }

    private static UpdateStartingCondition(kitty: Kitty) {
        const currentSafezone: number = kitty.CurrentSafeZone
        const kitties = Globals.ALL_KITTIES_LIST
        let skippedSafezone: boolean = false

        for (let i = 0; i < kitties.length; i++) {
            const currentKitty: Kitty = kitties[i]
            if (
                currentKitty.CurrentSafeZone !== currentSafezone - 1 &&
                currentKitty.CurrentSafeZone !== currentSafezone
            ) {
                skippedSafezone = true
                break
            }
        }

        if (skippedSafezone) {
            ChainedTogether.IsStartingContidionValid = false
        }
    }

    /// <summary>
    /// Starts the event
    /// </summary>
    public static StartEvent = () => {
        if (!ChainedTogether.EventTriggered) {
            ChainedTogether.IsStartingContidionValid = true
            return
        }

        ChainedTogether.EventStarted = true

        try {
            ChainedTogether.SetGroups()
            ChainedTogether.MoveChainTimer ??= Timer.create()
            ChainedTogether.MoveChainTimer.start(ChainedTogether.timerInterval, true, ChainedTogether.MoveChain)
        } catch (e) {
            Logger.Warning(`Error in ChainedTogether.StartEvent: ${e}`)
            throw e
        }
    }

    private static MoveChain = () => {
        const kitties = Globals.ALL_KITTIES_LIST
        let isOutOfRange: boolean = false
        let kittyOutOfRange: string = ''

        for (let i = 0; i < kitties.length - 1; i++) {
            const kitty = kitties[i]
            const kittyName = kitty.name

            if (!ChainedTogether.KittyLightnings.has(kittyName)) continue

            const chain = ChainedTogether.KittyLightnings.get(kittyName)!
            isOutOfRange = chain.Move()
            kittyOutOfRange = kittyName

            if (isOutOfRange) break
        }

        if (isOutOfRange) {
            ChainedTogether.LoseEvent(kittyOutOfRange)
        }
    }

    public static LoseEvent(kittyNameOutSideRange: string) {
        if (!ChainedTogether.EventStarted) return // Event not started or already ended.

        try {
            ChainedTogether.FreeKittiesFromGroup(kittyNameOutSideRange, false)
        } catch (e) {
            Logger.Warning(`Error in ChainedTogether.LoseEvent ${e}`)
            throw e
        }
    }

    /// <summary>
    /// Purpose is to regenerate the group chains if a player leaves the game or disconnects.
    /// </summary>
    /// <param name="kittyName"></param>
    public static RegenerateGroup(kittyName: string) {
        const groupIndex: number = ChainedTogether.kittyGroups.findIndex(group =>
            group.some(kitty => kitty.name === kittyName)
        ) // IEnumerable "Any" leaks
        if (groupIndex < 0) return

        try {
            const currentGroup = ChainedTogether.kittyGroups[groupIndex].filter(kitty => kitty.name !== kittyName) // Where and ToList are IEnumerable or Creating a new Object  .. LEAKS

            ChainedTogether.FreeKittiesFromGroup(kittyName, false)

            for (let j = 0; j < currentGroup.length - 1; j++) {
                const currentKitty = currentGroup[j]
                const nextKitty = currentGroup[j + 1]

                currentKitty.IsChained = true
                const chain: Chain = MemoryHandler.getEmptyObject<Chain>()
                chain.SetKitties(currentKitty, nextKitty)
                ChainedTogether.KittyLightnings.set(currentKitty.name, chain)
            }

            ChainedTogether.kittyGroups.push(currentGroup)
        } catch (e) {
            Logger.Warning(`Error in ChainedTogether.LoseEvent: ${e}`)
            throw e
        }
    }

    private static FreeKittiesFromGroup(kittyName: string, isVictory: boolean = false) {
        const groupIndex: number = ChainedTogether.kittyGroups.findIndex(group =>
            group.some(kitty => kitty.name === kittyName)
        ) //IEnumerable with "Any" leaks
        if (groupIndex < 0) return

        const currentGroup = ChainedTogether.kittyGroups[groupIndex]

        for (let i = 0; i < currentGroup.length; i++) {
            const kitty = currentGroup[i]
            kitty.IsChained = false

            if (isVictory) {
                ChainedTogether.AwardChainedTogether(kitty)
            }

            if (ChainedTogether.KittyLightnings.has(kitty.name)) {
                ChainedTogether.KittyLightnings.get(kitty.name)!.dispose()
                ChainedTogether.KittyLightnings.delete(kitty.name)
            }
        }

        ChainedTogether.kittyGroups.splice(groupIndex, 1)
    }

    private static SetGroups = () => {
        const allKitties = Globals.ALL_KITTIES_LIST
        const count: number = allKitties.length

        // Shuffle the kitties list to ensure randomness
        for (let i: number = allKitties.length - 1; i > 0; i--) {
            const j: number = Math.random()
            const temp: Kitty = allKitties[i]
            allKitties[i] = allKitties[j]
            allKitties[j] = temp
        }

        if (count < 3) {
            ChainedTogether.kittyGroups.push(allKitties)
            ChainedTogether.ChainGroup(allKitties)
        }

        let index = 0
        let groupsOfThree: number = count / 3
        let remainder: number = count % 3

        if (remainder === 1) {
            // convert two groups of 3 into two groups of 4 to avoid a group of 1
            groupsOfThree -= 1
            remainder += 3
        }

        for (let i = 0; i < groupsOfThree; i++) {
            const group = [allKitties[index], allKitties[index + 1], allKitties[index + 2]]
            ChainedTogether.kittyGroups.push(group)
            ChainedTogether.ChainGroup(group)
            index += 3
        }

        if (remainder > 0) {
            const lastGroup: Kitty[] = []
            for (let i: number = index; i < allKitties.length; i++) {
                lastGroup.push(allKitties[i])
            }
            ChainedTogether.kittyGroups.push(lastGroup)
            ChainedTogether.ChainGroup(lastGroup)
        }
    }

    private static ChainGroup(group: Kitty[]) {
        for (let j = 0; j < group.length - 1; j++) {
            const currentKitty = group[j]
            const nextKitty = group[j + 1]

            currentKitty.IsChained = true
            const chain: Chain = MemoryHandler.getEmptyObject<Chain>()
            chain.SetKitties(currentKitty, nextKitty)
            ChainedTogether.KittyLightnings.set(currentKitty.name, chain)
        }
    }

    public static ReachedSafezone(kitty: Kitty) {
        try {
            if (!ChainedTogether.EventStarted) {
                ChainedTogether.UpdateStartingCondition(kitty)

                if (ChainedTogether.IsInLastSafezone(kitty)) {
                    ChainedTogether.TriggerEventOnLastSafeZone()
                }
                return // Event not started or already ended.
            }

            if (!ChainedTogether.IsInLastSafezone(kitty)) return

            //finish event
            ChainedTogether.FreeKittiesFromGroup(kitty.name, true)
        } catch (e) {
            Logger.Warning(`Error in ChainedTogether.ReachedSafezone ${e}`)
            throw e
        }
    }

    private static AwardChainedTogether(kitty: Kitty) {
        Utility.CreateSimpleTextTag(`${Colors.COLOR_RED}Chained Together!${Colors.COLOR_RESET}`, 2.0, kitty.Unit)

        const level: DifficultyLevel = Difficulty.DifficultyValue

        const awards = Globals.GAME_AWARDS_SORTED.Auras
        let nameOfAward: string

        switch (level) {
            case DifficultyLevel.Nightmare:
                nameOfAward = 'ChainedNightmareAura'
                break
            case DifficultyLevel.Impossible:
                nameOfAward = 'ChainedImpossibleAura'
                break
            case DifficultyLevel.Hard:
                nameOfAward = 'ChainedHardAura'
                break
            default:
                nameOfAward = 'ChainedNormalAura'
                break
        }

        AwardManager.GiveReward(kitty.Player, nameOfAward)
    }

    private static IsInLastSafezone(kitty: Kitty) {
        // Check if the checks made with this function can be removed by calling the
        // functions TriggerEventOnLastSafeZone and ReachedSafezone on Source\Events\VictoryZone\FinalSafezone.cs
        // I'm afraid there could be a race condition by doing so
        return kitty.CurrentSafeZone === RegionList.SafeZones.length - 1
    }
}

export class Chain {
    public FirstKitty: Kitty
    public SecondKitty: Kitty
    public Lightning: lightning

    private static readonly ranges: { [key in DifficultyLevel]: { good: number; far: number; breakPoint: number } } = {
        [DifficultyLevel.Normal]: { good: 600, far: 700, breakPoint: 800 },
        [DifficultyLevel.Hard]: { good: 650, far: 750, breakPoint: 850 },
        [DifficultyLevel.Impossible]: { good: 700, far: 800, breakPoint: 900 },
        [DifficultyLevel.Nightmare]: { good: 750, far: 850, breakPoint: 950 },
    }

    public constructor() {}

    public SetKitties(firstKitty: Kitty, secondKitty: Kitty) {
        this.FirstKitty = firstKitty
        this.SecondKitty = secondKitty
        DestroyLightning(this.Lightning)
        this.Lightning = AddLightning(
            'WHCH',
            true,
            this.FirstKitty.Unit.x,
            this.FirstKitty.Unit.y,
            this.SecondKitty.Unit.x,
            this.SecondKitty.Unit.y
        )!
        this.FirstKitty.IsChained = true
        this.SecondKitty.IsChained = true
    }

    public Move(): boolean {
        const outOfRange: number = Chain.CalculateRangeByDifficulty('breakPoint')
        let isOutOfRange: boolean = false
        const x1 = this.FirstKitty.Unit.x
        const y1 = this.FirstKitty.Unit.y
        const x2 = this.SecondKitty.Unit.x
        const y2 = this.SecondKitty.Unit.y
        MoveLightning(
            this.Lightning,
            true,
            this.FirstKitty.Unit.x,
            this.FirstKitty.Unit.y,
            this.SecondKitty.Unit.x,
            this.SecondKitty.Unit.y
        )
        const distance: number = Math.abs(x2 - x1) + Math.abs(y2 - y1)

        if (distance > outOfRange) {
            isOutOfRange = true
        }

        this.ChangeChainColor(distance)
        return isOutOfRange
    }

    public dispose = () => {
        DestroyLightning(this.Lightning)
        this.FirstKitty.IsChained = false
        this.SecondKitty.IsChained = false
        MemoryHandler.destroyObject(this)
    }

    public ChangeChainColor(distance: number) {
        let red = 0.0
        let green = 1.0
        let blue = 0.0
        const alpha = 1.0 // Default color is green

        const far: number = Chain.CalculateRangeByDifficulty('far')
        const good: number = Chain.CalculateRangeByDifficulty('good')

        if (distance > far) {
            red = 1.0
            green = 0.0
            blue = 0.0 // Red
        } else if (distance > good) {
            red = 1.0
            green = 1.0
            blue = 0.0 // Yellow
        }

        SetLightningColor(this.Lightning, red, green, blue, alpha)
    }

    public static CalculateRangeByDifficulty(rangeType: string) {
        const level = Difficulty.DifficultyValue
        let selectedRange: { good: number; far: number; breakPoint: number }

        switch (true) {
            case level >= DifficultyLevel.Nightmare:
                selectedRange = Chain.ranges[DifficultyLevel.Nightmare]
                break
            case level >= DifficultyLevel.Impossible:
                selectedRange = Chain.ranges[DifficultyLevel.Impossible]
                break
            case level >= DifficultyLevel.Hard:
                selectedRange = Chain.ranges[DifficultyLevel.Hard]
                break
            default:
                selectedRange = Chain.ranges[DifficultyLevel.Normal]
                break
        }

        switch (rangeType) {
            case 'good':
                return selectedRange.good
            case 'far':
                return selectedRange.far
            case 'breakPoint':
                return selectedRange.breakPoint
            default:
                throw new Error(`Invalid rangeType '${rangeType}'`)
        }
    }
}
